### ISSUE TITLE
Provide custom openssl/curl version to cmake

### ISSUE THREAD

**Issue by aaronovz1** [aws/aws-sdk-cpp#938]
### What platform/OS are you using?
Ubuntu 16.04

### What compiler are you using? what version?
GCC 5.4

### What's your CMake arguments?
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX:PATH=/home/user/aws-sdk-shared -DOPENSSL_INCLUDE_DIR=/home/user/openssl1.0.2l/include -DOPENSSL_LIBRARIES="/home/user/openssl1.0.2l/libshared/libssl.so;/home/user/openssl1.0.2l/libshared/libcrypto.so" ..

We compile our applications against a different version than what is on the system path and with the default compilation of the AWS SDK I get linker errors when compiling our applications because there is a mismatch in symbols from the system version openssl and the one we compiled ourselves.

I tried to supply custom include/lib paths to cmake but it doesn't appear to work correctly. Is there a way to correctly bypass the FindOpenSSL cmake module and supply our own version of OpenSSL and CURL?

**Comment by crusader-mike** [comment]
One way is to install newer version of libcurl into /usr/local, e.g.:
```
cd ~
wget https://curl.haxx.se/download/curl-7.54.0.tar.gz
tar -xvf curl-7.54.0.tar.gz
cd curl-7.54.0
./configure --libdir=/usr/local/lib64
make -j
sudo make install
sudo ldconfig
```
and link it statically (to avoid problems during deployment):
```
cmake -DBUILD_SHARED_LIBS=0 ...
```

... or package libcurl with your app. I used static linking -- not ideal, but my app is used in controlled environment, no pressing need for security updates.

**Comment by marcomagdy** [comment]
Try passing -DCMAKE_PREFIX_PATH=<path to openssl where lib and include dirs are found>

**Comment by aaronovz1** [comment]
I had to use DOPENSSL_ROOT_DIR instead of specifying separate paths for includes and lib directories. There wasn't an equivalent variable for CURL so I had to hack up external_dependencies.cmake to specify a different path.

**Comment by hassanctech** [comment]
Use the following for CURL:  -DCURL_INCLUDE_DIR="${DEPEND_BUILD_DIR}/curl-7.57.0/include" -DCURL_LIBRARY="${DEPEND_BUILD_DIR}/curl-7.57.0/lib"

This is what we're using and it's works for us.  We also had to set the DOPENSSL_ROOT_DIR.  In addition we needed to set our own zlib as follows:  -DZLIB_ROOT="${DEPEND_BUILD_DIR}/zlib-1.2.11"

Before this we were also hacking the external_dependencies.cmake, but with the above you should not have to do that anymore.

**Comment by rvernica** [comment]
It might be because this thread is old, but the `CURL_INCLUDE_DIR` and `CURL_LIBRARY` are not supported in `aws-sdk-cpp-1.8.3`:
```
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CURL_INCLUDE_DIR
```
```
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CURL_LIBRARY
```

**Comment by ewan15** [comment]
@rvernica Did you find workaround to this problem?

**Comment by rvernica** [comment]
No, I tried various things but I could not make `cmake` to use my `cURL` build. I gave up on this and managed to get my application link against my `cURL` build. See [here](https://github.com/aws/aws-sdk-cpp/issues/1491#issuecomment-707472688) for my original issue and the workaround.

**Comment by ChuckCottrill** [comment]
After examining the aws-sdk cmake files, I discovered that the aws-sdk cmake provides support for both system installed CURL and local installed curl, but the system installed CURL takes precedence by using `include(FindCURL)` first.
I have found that there is a system-wide setting, `NO_CMAKE_SYSTEM_PATH`, but that prevents finding any system installed libraries. I was able to use `NO_CMAKE_SYSTEM_PATH` as a workaround, but that would prevent use of other system libraries.
